### PR TITLE
Force the scan-build workflow to fail if any bugs found

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
As far as I noticed, currently the scan-build workflow doesn't fail if any bugs are found. This PR fixes this. Also it fixes the following false-positives (reported also by SonarCloud):

https://sonarcloud.io/project/issues?id=ihhub_fheroes2&open=AXhxYwUmQocKC9FrtX__&resolved=false&rules=cpp%3AS2259&types=BUG
https://sonarcloud.io/project/issues?id=ihhub_fheroes2&open=AXhxYwUmQocKC9FrtYAA&resolved=false&rules=cpp%3AS2259&types=BUG

As far as I can tell, there is no way that `hero` and `castle` can somehow become `nullptr` here, so I added `assert()` calls to convince scan-build of this.

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
Hi @ihhub I was thinking of doing a similar PR for cppclean (static_analysis.yml), but it turns out that due to specific structure of this project it reports a lots of false positives (especially "declared but not defined" and "already included"). Is it worth having a check that no one is looking at the results of?

**Comment by ihhub** [comment]
> Hi @ihhub I was thinking of doing a similar PR for cppclean (static_analysis.yml), but it turns out that due to specific structure of this project it reports a lots of false positives (especially "declared but not defined" and "already included"). Is it worth having a check that no one is looking at the results of?

cppclean was useful at the beginning of the project. Right now it doesn't have any value. You could create a pull request to remove the job.

File: .github/workflows/scan-build.yml [hunk 1]
```diff
@@ -18,8 +18,9 @@ jobs:
         sudo apt-get update
         sudo apt-get install -y libsdl2-dev libsdl2-ttf-dev libsdl2-mixer-dev libsdl2-image-dev gettext clang-tools
     - name: compile
-      run: scan-build -v make -j 2
+      run: scan-build --status-bugs -v make -j 2
       env:
+        DEBUG: "ON"
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply .github/workflows/scan-build.yml:23]: Hi @ihhub also to verify code that is used only in debug builds.

[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply .github/workflows/scan-build.yml:23]: @ihhub BTW I would like to propose building all pull_request workflows with DEBUG set. I've had cases where I forgot to update variables in DEBUG_LOG calls and haven't noticed this in regular build.

[the remark this answers is not in the document]
Comment (ihhub) [reply .github/workflows/scan-build.yml:23]: We could build a separate debug build for this. If we build all debug builds then we could have a situation when a code for release version is broken but we don't notice it.

[the remark this answers is not in the document]
Comment (ihhub) [reply .github/workflows/scan-build.yml:23]: I don't think that SonarQube should be aware of these assertions. After all it's obvious that this is a false positive case which we could manually remove from SonarQube reports.

[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply .github/workflows/scan-build.yml:23]: Hi @ihhub Those assertions are for scan-build, not for SonarCloud per se. SonarCloud simply uses scan-build just like this workflow :) So those asserts will fix both at once.

[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply .github/workflows/scan-build.yml:23]: @ihhub May be then add targets with DEBUG set at least in Linux workflow? Let's say, sdl-debug and sdl2-debug. A can make it as a separate PR.

[the remark this answers is not in the document]
Comment (ihhub) [reply .github/workflows/scan-build.yml:23]: @oleg-derevenetz , let's do this way and only for pull requests.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub I also made a separate scan-build jobs with and without DEBUG. Or maybe do a scan-build only for pull requests as well?

**Comment by ihhub** [comment]
> Hi @ihhub I also made a separate scan-build jobs with and without DEBUG. Or maybe do a scan-build only for pull requests as well?

We should limit the number of runners for each commit as GitHub Actions has a limitation on resources. Is there a need to run scan-build with debug mode except compilation error catches?

**Comment by oleg-derevenetz** [comment]
@ihhub

> Is there a need to run scan-build with debug mode except compilation error catches?

Well, in fact, it's not necessary, if we already have debug builds in `linux_pull_request.yml`. I'll rework this to run scan-build only on release build.

**Comment by ihhub** [comment]
@oleg-derevenetz , excellent work!

**Pull request merged**